File: src/bet/sizing/kelly.py
"""Kelly Criterion bet sizer."""

from __future__ import annotations

from .types import BetSizing, KellyFraction


def full_kelly(edge: float, decimal_odds: float) -> float:
    """Compute the full Kelly fraction.

    f* = max(0, edge / (decimal_odds − 1))

    Args:
        edge: EV per unit staked = model_prob × decimal_odds − 1.
        decimal_odds: Offered decimal odds (must be > 1.0 to be meaningful).

    Returns:
        Optimal bankroll fraction to wager. 0.0 when edge ≤ 0 or odds ≤ 1.
    """
    if decimal_odds <= 1.0:
        return 0.0
    return max(0.0, edge / (decimal_odds - 1.0))


def compute_kelly_fraction(edge: float, decimal_odds: float) -> KellyFraction:
    """Return all three Kelly variants for the given edge and odds.

    Args:
        edge: EV per unit staked.
        decimal_odds: Offered decimal odds.

    Returns:
        KellyFraction with full, half, and quarter Kelly values.
    """
    f = full_kelly(edge, decimal_odds)
    return KellyFraction(
        full_kelly=f,
        half_kelly=f / 2.0,
        quarter_kelly=f / 4.0,
        edge=edge,
        decimal_odds=decimal_odds,
    )


class KellySizer:
    """Sizes bets using a configurable fraction of the Kelly Criterion.

    Args:
        fraction: Multiplier applied to full Kelly. 1.0 = full Kelly,
            0.5 = half Kelly, 0.25 = quarter Kelly. Defaults to 1.0.
    """

    def __init__(self, fraction: float = 1.0) -> None:
        self._fraction = fraction

    def size(self, edge: float, decimal_odds: float, bankroll: float) -> BetSizing:
        """Compute Kelly-sized stake.

        Args:
            edge: EV per unit staked.
            decimal_odds: Offered decimal odds.
            bankroll: Current bankroll.

        Returns:
            BetSizing with Kelly-derived stake, never negative.
        """
        f = full_kelly(edge, decimal_odds) * self._fraction
        return BetSizing(stake=bankroll * f, fraction=f, bankroll=bankroll)

File: internal/sizing/kelly.go
package sizing

import "math"

// FullKelly returns f* = max(0, edge / (decimalOdds − 1)).
//
// Returns 0 when edge ≤ 0 or decimalOdds ≤ 1.
func FullKelly(edge, decimalOdds float64) float64 {
	if decimalOdds <= 1.0 {
		return 0.0
	}
	return math.Max(0.0, edge/(decimalOdds-1.0))
}

// ComputeKellyFraction returns all three Kelly variants for the given edge and odds.
func ComputeKellyFraction(edge, decimalOdds float64) KellyFraction {
	f := FullKelly(edge, decimalOdds)
	return KellyFraction{
		Full:    f,
		Half:    f / 2.0,
		Quarter: f / 4.0,
	}
}

// KellySizer sizes bets using a configurable fraction of the Kelly Criterion.
type KellySizer struct {
	// Fraction is the multiplier applied to full Kelly.
	// 1.0 = full Kelly, 0.5 = half Kelly, 0.25 = quarter Kelly.
	Fraction float64
}

// Size computes the Kelly-derived stake.
func (s KellySizer) Size(edge, decimalOdds, bankroll float64) BetSizing {
	f := FullKelly(edge, decimalOdds) * s.Fraction
	return BetSizing{
		Stake:    bankroll * f,
		Fraction: f,
		Bankroll: bankroll,
	}
}
